README.md
<p>
  <a href="https://github.com/oheil/NormalizeQuantiles.jl/actions">
    <img src="https://img.shields.io/github/workflow/status/oheil/NormalizeQuantiles.jl/Mac%20OS%20X?label=OSX&logo=Apple&logoColor=white&style=flat-square">
  </a>
  <a href="https://github.com/oheil/NormalizeQuantiles.jl/actions">
    <img src="https://img.shields.io/github/workflow/status/oheil/NormalizeQuantiles.jl/Linux?label=Linux&logo=Linux&logoColor=white&style=flat-square">
  </a>
  <a href="https://github.com/oheil/NormalizeQuantiles.jl/actions">
    <img src="https://img.shields.io/github/workflow/status/oheil/NormalizeQuantiles.jl/Windows?label=Windows&logo=Windows&logoColor=white&style=flat-square">
  </a>
  <a href="https://codecov.io/gh/oheil/NormalizeQuantiles.jl">
    <img src="https://img.shields.io/codecov/c/github/oheil/NormalizeQuantiles.jl/master?label=Codecov&logo=codecov&logoColor=white&style=flat-square">
  </a>
</p>

[pkgeval-img]: https://juliaci.github.io/NanosoldierReports/pkgeval_badges/N/NormalizeQuantiles.svg
[pkgeval-url]: https://juliaci.github.io/NanosoldierReports/pkgeval_badges/report.html

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![PkgEval][pkgeval-img]][pkgeval-url]

[![deps](https://juliahub.com/docs/NormalizeQuantiles/deps.svg)](https://juliahub.com/ui/Packages/NormalizeQuantiles/D7PQ0?t=2)
[![version](https://juliahub.com/docs/NormalizeQuantiles/version.svg)](https://juliahub.com/ui/Packages/NormalizeQuantiles/D7PQ0)
[![pkgeval](https://juliahub.com/docs/NormalizeQuantiles/pkgeval.svg)](https://juliahub.com/ui/Packages/NormalizeQuantiles/D7PQ0)

# NormalizeQuantiles

For julia 0.4, 0.5, 0.6 see: https://github.com/oheil/NormalizeQuantiles.jl/tree/backport-0.6

Package NormalizeQuantiles implements quantile normalization

```julia
qn = normalizeQuantiles(array);
```

and provides a function to calculate sample ranks

```julia
(r,m) = sampleRanks(array);
```

of a given vector or matrix.

**References**

* Amaratunga, D.; Cabrera, J. (2001). "Analysis of Data from Viral DNA Microchips". Journal of the American Statistical Association. 96 (456): 1161. [doi:10.1198/016214501753381814](https://doi.org/10.1198/016214501753381814)
* Bolstad, B. M.; Irizarry, R. A.; Astrand, M.; Speed, T. P. (2003). "A comparison of normalization methods for high density oligonucleotide array data based on variance and bias". Bioinformatics. 19 (2): 185–193. [doi:10.1093/bioinformatics/19.2.185](https://doi.org/10.1093/bioinformatics/19.2.185) [PMID 12538238](https://www.ncbi.nlm.nih.gov/pubmed/12538238)
* Wikipedia contributors. (2018, June 12). Quantile normalization. In Wikipedia, The Free Encyclopedia. Retrieved 11:54, August 3, 2018, from [https://en.wikipedia.org/w/index.php?title=Quantile_normalization](https://en.wikipedia.org/w/index.php?title=Quantile_normalization)

**Table of Contents**

* [Dependencies](#dependencies)
* [Remarks](#remarks)
* [Usage examples `normalizeQuantiles`](#usage-examples-normalizequantiles)
  * [General usage](#general-usage)
  * [Missing Values](#missing-values)
  * [SharedArray and multicore usage examples](#sharedarray-and-multicore-usage-examples)
  * [OffsetArrays](#offsetarrays)
* [Behaviour of function `normalizeQuantiles`](#behaviour-of-function-normalizequantiles)
* [Data prerequisites](#data-prerequisites)
* [Remarks on data with missing values](#remarks-on-data-with-missing-values)
* [List of all exported definitions for `normalizeQuantiles`](#list-of-all-exported-definitions-for-normalizequantiles)
* [Usage examples `sampleRanks`](#usage-examples-sampleranks)
* [List of all exported definitions for `sampleRanks`](#list-of-all-exported-definitions-for-sampleranks)

## Dependencies

#### Julia versions

* Julia 0.7 or above

#### Third party packages

* none

#### Standard Library packages

* [Distributed](https://docs.julialang.org/en/v1/stdlib/Distributed/)
* [SharedArrays](https://docs.julialang.org/en/v1/stdlib/SharedArrays/)
* [Random](https://docs.julialang.org/en/v1/stdlib/Random/)
* [Statistics](https://docs.julialang.org/en/v1/stdlib/Statistics/)

## Remarks

* for julia 0.4, 0.5, 0.6 see: https://github.com/oheil/NormalizeQuantiles.jl/tree/backport-0.6
* Code examples and output on this page have been used on and copied from the julia 0.7 [REPL](https://docs.julialang.org/en/latest/manual/interacting-with-julia/)
* Last commit with julia 0.3 support: [Jan 20, 2017, eb97d24ff77d470d0d121fabf83d59979ad0db36](https://github.com/oheil/NormalizeQuantiles.jl/tree/eb97d24ff77d470d0d121fabf83d59979ad0db36)
  * git checkout eb97d24ff77d470d0d121fabf83d59979ad0db36

## Usage examples `normalizeQuantiles`

#### General usage
 
```julia
Pkg.add("NormalizeQuantiles");
using NormalizeQuantiles;
```

The following `array` is interpreted as a matrix with 4 rows and 3 columns:

```julia
array = [ 3.0 2.0 1.0 ; 4.0 5.0 6.0 ; 9.0 7.0 8.0 ; 5.0 2.0 8.0 ];
qn = normalizeQuantiles(array)
```
```
	julia> qn
	4×3 Array{Float64,2}:
	 2.0  3.0  2.0
	 4.0  6.0  4.0
	 8.0  8.0  7.0
	 6.0  3.0  7.0
```

The columns in `qn` are now quantile normalized to each other.

The input array must not have dimension larger than 2.

Return type of function normalizeQuantiles is always Array{Float64,2}

#### Missing Values

If your data contain some missing values like `NaN` (Not a Number) or something else, they will be changed to `NaN`:

```julia
array = [ NaN 2.0 1.0 ; 4.0 "empty" 6.0 ; 9.0 7.0 8.0 ; 5.0 2.0 8.0 ];
```
```
	julia> array
	4×3 Array{Any,2}:
	 NaN    2.0       1.0
	4.0   "empty"  6.0
	9.0  7.0       8.0
	5.0  2.0       8.0
```
```julia
qn = normalizeQuantiles(array)
```
```
	julia> qn
	4×3 Array{Float64,2}:
	 NaN      3.25  1.5
	   5.0  NaN     5.0
	   8.0    8.0   6.5
	   5.0    3.25  6.5
```

NaN is of type Float64, so there is nothing similar for Int types.

```
	julia> typeof(NaN)
	Float64
```

You can convert the result to `Array{Union{Missing, Float64},2}` with:

```julia
qnMissing = convert(Array{Union{Missing,Float64}},qn)
```
```
	julia> qnMissing
	4×3 Array{Union{Missing, Float64},2}:
	 NaN      3.25  1.5
	   5.0  NaN     5.0
	   8.0    8.0   6.5
	   5.0    3.25  6.5
```
```julia
qnMissing[isnan.(qnMissing)] = missing;
```
```
	julia> qnMissing
	4×3 Array{Union{Missing, Float64},2}:
	  missing  3.25      1.5
	 5.0        missing  5.0
	 8.0       8.0       6.5
	 5.0       3.25      6.5
```

#### SharedArray and multicore usage examples

> Remark: restart julia now. `addprocs()` must be called before `using NormalizeQuantiles;`.

To use multiple cores on a single machine you can use the standard packages `Distributed` and `SharedArrays`:

```julia
using Distributed
addprocs();
@everywhere using SharedArrays
@everywhere using NormalizeQuantiles

sa = SharedArray{Float64}([ 3.0 2.0 1.0 ; 4.0 5.0 6.0 ; 9.0 7.0 8.0 ; 5.0 2.0 8.0 ])
```
```
	julia> sa
	4×3 SharedArray{Float64,2}:
	 3.0  2.0  1.0
	 4.0  5.0  6.0
	 9.0  7.0  8.0
	 5.0  2.0  8.0
```
```julia
qn = normalizeQuantiles(sa)
```
```
	julia> qn
	4×3 Array{Float64,2}:
	 2.0  3.0  2.0
	 4.0  6.0  4.0
	 8.0  8.0  7.0
	 6.0  3.0  7.0
```

> Remark: restart julia again.

For small data sets using `Distributed` and `SharedArrays` decreases performance:

```julia
using NormalizeQuantiles
la = randn((100,100));
normalizeQuantiles(la); @time normalizeQuantiles(la);
```
```
	julia> @time normalizeQuantiles(la);
	  0.003178 seconds (8.35 k allocations: 2.813 MiB)
```

> Remark: restart julia.

```julia
using Distributed
addprocs();
@everywhere using SharedArrays
@everywhere using NormalizeQuantiles
sa = SharedArray{Float64}( randn((100,100)) );
normalizeQuantiles(sa); @time normalizeQuantiles(sa);
```
```
	julia> @time normalizeQuantiles(sa);
	  0.013014 seconds (12.10 k allocations: 432.146 KiB)
```

> Remark: restart julia.

For larger data sets performance increases with multicore processors:

```julia
using NormalizeQuantiles
la = randn((1000,10000));
normalizeQuantiles(la); @time normalizeQuantiles(la);
```
```
	julia> @time normalizeQuantiles(la);
	  2.959431 seconds (784.18 k allocations: 2.281 GiB, 12.13% gc time)
```

> Remark: restart julia.

```julia
using Distributed
addprocs();
@everywhere using SharedArrays
@everywhere using NormalizeQuantiles
la = randn((1000,10000));
sa = SharedArray{Float64}(la);
normalizeQuantiles(sa); @time normalizeQuantiles(sa);
```
```
	julia> @time normalizeQuantiles(sa);
	  1.030016 seconds (83.85 k allocations: 80.754 MiB, 5.77% gc time)
```
Using non-SharedArrays in a multicore setup is slowest:
```
	julia> normalizeQuantiles(la); @time normalizeQuantiles(la);
	  5.776685 seconds (294.06 k allocations: 92.532 MiB, 0.28% gc time)
```

#### OffsetArrays

> Remark: with Julia 1.3.1 OffsetArrays are not supported until https://github.com/JuliaLang/julia/pull/34886 is released (expected in Julia 1.5)

```
using NormalizeQuantiles, OffsetArrays

array = [ 3 missing 1 ; 4 5 6 ; missing 7 8 ; 5 2 8 ];
oa = OffsetArray(array,-1,-1);
```
```
julia> oa
4×3 OffsetArray(::Array{Union{Missing, Int64},2}, 0:3, 0:2) with eltype Union{Missing, Int64} with indices 0:3×0:2:
 3          missing  1
 4         5         6
  missing  7         8
 5         2         8
```

```
qn = normalizeQuantiles(oa);
```
The quantile normalized result is not an OffsetArray:
```
julia> qn
4×3 Array{Float64,2}:
   2.0      NaN        2.0
   4.0        6.5      4.0
 NaN          6.66667  6.58333
   6.66667    4.0      6.58333
```

## Behaviour of function `normalizeQuantiles`

After quantile normalization the sets of values of each column have the same statistical properties.
This is quantile normalization without a reference column.

The function `normalizeQuantiles` expects an array with dimension <= 2 and always returns a matrix of same dimensions as the input matrix and of type `Array{Float64,2}`.

`NaN` values of type `Float64` and any other non-numbers, like strings, are treated as random missing values and the result value will be `NaN`. See "Remarks on data with missing values" below.

Equal values in a column of the input matrix will have different quantile normalized values. Those different result values can't be assigned back to the proper original positions because they are indistinguishable. The mean value of the different result values are therefor put back into original positions.

Example:
```
julia> array = [ 1 2 ; 2  3 ; 2 5 ]
3×2 Matrix{Int64}:
 1  2
 2  3
 2  5

julia> qn = normalizeQuantiles(array)
3×2 Matrix{Float64}:
 1.5  1.5
 3.0  2.5
 3.0  3.5
```
In row 2 and 3 instead if `2.5` and `3.5` the mean `3.0` is the result in both rows.

## Data prerequisites

To use quantile normalization your data should have the following properties:

* the distribution of values in each column should be similar
* number of values for each column should be large
* number of missing values in the data should be small and of random nature

## Remarks on data with missing values

Currently there seems to be no general agreement on how to deal with missing values during quantile normalization. Here we put any given missing value back into the sorted column at the original position before calculating the mean of the rows.

## List of all exported definitions for `normalizeQuantiles`

| | normalizeQuantiles |
| -----------------------: | ----------------------- | 
| **Definition:** | `Array{Float64,2} function normalizeQuantiles(matrix::AbstractArray)` |
| Input type: | `matrix::AbstractArray` |
| Return type: | `Array{Float64,2}` |


## Usage examples `sampleRanks`

`sampleRanks` of a given vector calculates for each element the rank, which is the position of the element in the sorted vector.

```julia
using NormalizeQuantiles
a = [ 5.0 2.0 4.0 3.0 1.0 ];
(r,m) = sampleRanks(a);   # here only return value r is relevant, for m see below
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	 5
	 2
	 4
	 3
	 1
```

If you provide a matrix like

```julia
array = [ 1.0 2.0 3.0 ; 4.0 5.0 6.0 ; 7.0 8.0 9.0 ; 10.0 11.0 12.0 ]
```
```
	julia> array
	4×3 Array{Float64,2}:
	  1.0   2.0   3.0
	  4.0   5.0   6.0
	  7.0   8.0   9.0
	 10.0  11.0  12.0
```

ranks are calculated column wise, or in other words, array is treated as `array[:]`:
```julia
(r,m) = sampleRanks(array);
r
```
```
	julia> r
	12-element Array{Union{Missing, Int64},1}:
	  1
	  4
	  7
	 10
	  2
	  5
	  8
	 11
	  3
	  6
	  9
	 12
```

There are three optional keyword parameters `tiesMethod`, `naIncreasesRank` and `resultMatrix`:

```julia
(r,m) = sampleRanks(a,tiesMethod=tmMin,naIncreasesRank=false,resultMatrix=true);
(r,m) = sampleRanks(a,resultMatrix=true);
```

Equal values in the vector are called ties. There are several methods available on how to treat ties:
* tmMin : the smallest rank for all ties (default)
* tmMax : the largest rank
* tmOrder : increasing ranks
* tmReverse : decreasing ranks
* tmRandom : the ranks are randomly distributed
* tmAverage : the average rounded to the next integer

These methods are defined and exported as
	
```julia
	@enum qnTiesMethods tmMin tmMax tmOrder tmReverse tmRandom tmAverage
```

Internally ties have increasing ranks. On these the chosen method is applied.

The next rank for the successive values after the ties is the so far highest used rank plus 1.
	
Examples:

```julia
a = [ 7.0 2.0 4.0 2.0 1.0 ];
(r,m) = sampleRanks(a); #which is the same as (r,m)=sampleRanks(a,tiesMethod=tmMin)
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	 4
	 2
	 3
	 2
	 1
```
```julia
(r,m) = sampleRanks(a,tiesMethod=tmMax);
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	 5
	 3
	 4
	 3
	 1
```
```julia
(r,m) = sampleRanks(a,tiesMethod=tmReverse);
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	 5
	 3
	 4
	 2
	 1
```

One or more missing values in the vector are never equal and remain on there position after sorting. The rank of each missing value is always missing::Missing. The default is that a missing value does not increase the rank for successive values. Giving true keyword parameter `naIncreasesRank` changes that behavior to increasing the rank by 1 for successive values:

```julia
a = [ 7.0 2.0 4.0 2.0 1.0 ];
a[1] = NaN;
(r,m) = sampleRanks(a);
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	  missing
	 2
	 3
	 2
	 1
```
```julia
(r,m) = sampleRanks(a,naIncreasesRank=true);
r
```
```
	julia> r
	5-element Array{Union{Missing, Int64},1}:
	  missing
	 3
	 4
	 3
	 2
```

The keyword parameter `resultMatrix` lets you generate a dictionary of rank indices to allow direct access to all values with a given rank. For large vectors this may have a large memory consumption therefor the default is to return an empty dictionary of type `Dict{Int64,Array{Int64,N}}`:

```julia
a = [ 7.0 2.0 4.0 2.0 1.0 ];
(r,m) = sampleRanks(a,resultMatrix=true);
m
```
```
	julia> m
	Dict{Int64,Array{Int64,N} where N} with 4 entries:
	  4 => [1]
	  2 => [2,4]
	  3 => [3]
	  1 => [5]
```
```julia
haskey(m,2)   #does rank 2 exist?
```
```
	julia> haskey(m,2)
	true
```
```julia
a[m[2]]   #all values of rank 2
```
```
	julia> a[m[2]]
	2-element Array{Float64,1}:
	 2.0
	 2.0
```

## List of all exported definitions for `sampleRanks`

| | sampleRanks |
| -----------------------: | ----------------------- | 
| **Definition:** | `@enum qnTiesMethods tmMin tmMax tmOrder tmReverse tmRandom tmAverage` |
| Description: ||
| tmMin | the smallest rank for all ties |
| tmMax | the largest rank |
| tmOrder | increasing ranks |
| tmReverse | decreasing ranks |
| tmRandom | the ranks are randomly distributed |
| tmAverage | the average rounded to the next integer |

| | sampleRanks | |
| -----------------------: | ----------------------- | ----------------------- | 
| **Definition:** | `(Array{Union{Missing,Int},1},Dict{Int,Array{Int}}) sampleRanks(array::AbstractArray; tiesMethod::qnTiesMethods=tmMin, naIncreasesRank=false, resultMatrix=false)` | **keyword arguments** |
| Input type: | `array::AbstractArray` | data |
| Input type: | `tiesMethod::qnTiesMethods` | how to treat ties (default: `tmMin`) |
| Input type: | `naIncreasesRank::bool` | increase rank by one if NA (default: `false`) |
| Input type: | `resultMatrix::bool` | create rank dictionary (default: `false`) |
| Return type: | `(Array{Union{Missing,Int},1},Dict{Int,Array{Int}})` ||




CHANGELOG.md
## 1.2.2 (2022-02-14)

#### Features:

* none
	
#### Bug fixes:

* SharedArray/Distributed bug 

#### Remarks:

* none

## 1.2.1 (2021-04-16)

#### Features:

* OffsetArrays tests added
	
#### Bug fixes:

* none

#### Remarks:

* none

## 1.2.0 (2020-02-08)

#### Features:

* OffsetArrays now possible
	
#### Bug fixes:

* Array/Matrix of Int with missings problem resolved

#### Remarks:

  * with Julia 1.3 OffsetArrays are not working, need to wait until https://github.com/JuliaLang/julia/pull/34886) is released

## 0.3.3 (2017-12-20)

#### Features:

* none
	
#### Bug fixes:

* none

#### Remarks:

  * julia 0.7-DEV introduces "missing", therefor major recoding is needed.
  * This version will be freezed for julia 0.4 to 0.6(including)

## 0.3.2

#### Features:

* none
	
#### Bug fixes:

* none

#### Remarks:

  * Julia 0.7-DEV sharedArrayes removed from base

## 0.3.1 (2017-05-16)

#### Features:

* none
	
#### Bug fixes:

* none

#### Remarks:

  * Julia 0.6 syntax change and deprecation
  
    * WARNING: !(A::AbstractArray{Bool}) is deprecated, use .!(A) instead.

## 0.3.0 (2017-01-20)

#### Features:

  - none
	
#### Bug fixes:

  - none

#### Remarks:

  - dropped julia 0.3 support

## 0.2.1 (2017-01-20)

#### Features:

  - none
	
#### Bug fixes:

  - none

#### Remarks:

  - Only changes to reflect changes in syntax coming with julia 0.6
  - This version still supports julia 0.3

## 0.2.0 (2016-08-11)

#### Features:

  - started changelog
  - for function `sampleRanks` removed methods with optional parameters in favor of methods with keyword parameters
	
#### Bug fixes:

  - none

#### Remarks:

In julia 0.5 and 0.6 using optional parameters in functions like

```julia
function test(a,b=1,c=1)
	a+b+c
end


```

together with keyword parameters:

```julia
function test(a;b=1,c=1)
	a+b+c
end


```

results in warning messages when importing the package with `using`:

```
WARNING: Method definition test(Any) in ... overwritten at ...
```
	





src/NormalizeQuantiles.jl
module NormalizeQuantiles

export normalizeQuantiles

export sampleRanks
export qnTiesMethods,tmMin,tmMax,tmOrder,tmReverse,tmRandom,tmAverage

using Distributed
using SharedArrays
using Random
using Statistics

@enum qnTiesMethods tmMin tmMax tmOrder tmReverse tmRandom tmAverage

function checkForNotANumber(x::Any)
    (!isa(x,Integer) && !isa(x,Real)) || isnan(x)
end

@doc "
### qnmatrix::Array{Float64} function normalizeQuantiles(matrix::AbstractArray)
Calculate the quantile normalized data for the input matrix

Parameter:
    matrix::AbstractArray
The input data as an array of values interpreted as matrix(rows,columns)

Return value: 
    qnmatrix::Array{Float64}
The quantile normalized data as Array{Float64}

Example:

    using NormalizeQuantiles

    array = [ 3.0 2.0 1.0 ; 4.0 5.0 6.0 ; 9.0 7.0 8.0 ; 5.0 2.0 8.0 ]
    qn = normalizeQuantiles(array)

    row = 2
    column = 2
    array=convert(Array{Any},array)
    array[row,column] = missing
    qn = normalizeQuantiles(array)
"
function normalizeQuantiles(matrix::AbstractArray)
    if ndims(matrix) > 2
        throw(ArgumentError("normalizeQuantiles expects an array of dimension 2"))
    end
    nrows=size(matrix,1)
    ncols=size(matrix,2)
    # preparing the result matrix
    qnmatrix=SharedArray{Float64}(nrows,ncols)
    if ncols>0 && nrows>0
        # foreach column: sort the values without NAs; put NAs (if any) back into sorted list
        NormalizeQuantiles.sortColumns!(matrix,qnmatrix)
        # foreach row: set all values to the mean of the row, except NAs
        NormalizeQuantiles.meanRows!(qnmatrix)
        # foreach column: equal values in original column should all be mean of normalized values
        # foreach column: reorder the values back to the original order
        NormalizeQuantiles.equalValuesInColumnAndOrderToOriginal!(matrix,qnmatrix,nrows)
    end
    convert(Array{Float64},qnmatrix)
end

function sortColumns!(matrix::AbstractArray,qnmatrix::SharedArray{Float64})
    ncols=size(matrix,2)
    tcols=1:ncols
    scolumns=collect(eachindex(matrix[end,:]))
    @inbounds @sync @distributed for tcol in tcols
        scolumn=scolumns[tcol]
        sortcol=[ NormalizeQuantiles.checkForNotANumber(x) ? NaN : Float64(x) for x in matrix[:,scolumn] ]
        missingIndices=findall(isnan.(sortcol))
        sort!(sortcol)
        #putting original NaNs back into place:
        nMissPos=lastindex(missingIndices)
        delta=nMissPos
        index=lastindex(sortcol)
        while nMissPos>0
            missingPos=missingIndices[nMissPos]
            if index==missingPos
                sortcol[index]=NaN
                nMissPos-=1
                delta-=1
            else
                sortcol[index]=sortcol[index-delta]
            end
            index-=1
        end
        qnmatrix[:,tcol]=sortcol
    end
end

function meanRows!(qnmatrix::SharedArray{Float64})
    @inbounds @sync @distributed for srow = eachindex(qnmatrix[:,end])
        goodIndices=.!isnan.(qnmatrix[srow,:])
        rowView=view(qnmatrix,srow,goodIndices)      
        #rowmean=mean(rowView)
        #
        #in julia function sum the order of summing up the elements can vary on different
        #types of AbstractArray (e.g. qnmatrix[row,goodIndices] and rowView). Summing over
        #a large number of floats in different order yield different results because of 
        #floating point precision, so we calculate the sum in a well defined way:
        lrow=length(rowView)
        sum=0.0
        for i in eachindex(rowView)
            sum+=rowView[i]
        end
        rowmean=sum/lrow
        rowView.=rowmean
    end
end

function equalValuesInColumnAndOrderToOriginal!(matrix::AbstractArray,qnmatrix::SharedArray{Float64},nrows)
    ncols=size(matrix,2)
    tcols=1:ncols
    scolumns=collect(eachindex(matrix[end,:]))
    @inbounds @sync @distributed for tcol in tcols
        scolumn=scolumns[tcol]
        goodIndices=.!isnan.(qnmatrix[:,tcol])
        #matrix can be an OffsetArray
        colview=view(matrix,collect(eachindex(matrix[:,scolumn]))[goodIndices],scolumn)
        sortp=sortperm(Float64.(colview))
        if length(sortp)>0
            NormalizeQuantiles.setMeanForEqualOrigValues!(colview[sortp],qnmatrix,tcol,goodIndices)
        end
        qnmatrix[(1:nrows)[goodIndices][sortp],tcol]=qnmatrix[goodIndices,tcol]
    end
end

function setMeanForEqualOrigValues!(sortedArrayNoNAs::AbstractArray,qnmatrix::SharedArray{Float64},column::Int,goodIndices)
    nrows=length(sortedArrayNoNAs)
    foundIndices=zeros(Int,nrows)
    goodIndices2=findall(goodIndices)
    count=1
    lastValue=sortedArrayNoNAs[1]
    for i in 2:nrows
        nextValue=sortedArrayNoNAs[i]
        foundIndices[count]=goodIndices2[i-1]
        if nextValue==lastValue
            count+=1
        else
            if count>1
                qnmatrix[foundIndices[1:count],column].=mean(qnmatrix[foundIndices[1:count],column])
                count=1
            end
        end
        lastValue=nextValue
    end
    foundIndices[count]=goodIndices2[nrows]
    if count>1
        qnmatrix[foundIndices[1:count],column].=mean(qnmatrix[foundIndices[1:count],column])
    end
end

@doc "
### (Array{Union{Missing,Int}},Dict{Int,Array{Int}}) sampleRanks(array::AbstractArray;tiesMethod::qnTiesMethods=tmMin,naIncreasesRank=false,resultMatrix=false)

Calculate ranks of the values of a given vector.

Parameters:

    array: the input array
    tiesMethod: the method how ties (equal values) are treated
       possible values: tmMin tmMax tmOrder tmReverse tmRandom tmAverage
       default is tmMin
    naIncreasesRank: if true than any NA increases the following ranks by 1
    resultMatrix: if true than return a dictionary of rank keys and array of indices values

Example:
    
    using NormalizeQuantiles
    
    a = [ 5.0 2.0 4.0 3.0 1.0 ]
    
    (r,m)=sampleRanks(a)

    (r,m)=sampleRanks(a,tiesMethod=tmMin,naIncreasesRank=false,resultMatrix=true)
    
r is the vector of ranks.

m is a dictionary with rank as keys and as value the indices of all values of this rank.

"
function sampleRanks(array::AbstractArray;tiesMethod::qnTiesMethods=tmMin,naIncreasesRank::Bool=false,resultMatrix::Bool=false)
    nrows=length(array)
    naCounts=zeros(Int,nrows)
    goodIndices=falses(nrows)
    naCount=0
    reducedIndex=1
    goodIndex=1
    for arrayIndex in eachindex(array)
        if NormalizeQuantiles.checkForNotANumber(array[arrayIndex])
            naCount+=1
        else
            goodIndices[goodIndex]=true
            naCounts[reducedIndex]=naCount
            reducedIndex+=1
        end
        goodIndex+=1
    end
    reducedArraySorted=Float64.(array[goodIndices])
    reducedArraySortedIndices=sortperm(reducedArraySorted)
    reducedArraySorted=reducedArraySorted[reducedArraySortedIndices]
    result=Array{Union{Missing,Int}}(missing,nrows)
    rankMatrix=Dict{Int,Array{Int}}()
    group = Array{Int,1}()
    firstFound = true
    lastFound = NaN
    nextRank=1
    rankIncrement=0
    doIncreaseRank=false
    headingNA=false
    groupIndex=1
    offset=firstindex(array)-groupIndex
    for resultIndex in eachindex(array)
        if goodIndices[resultIndex]
            if headingNA
                headingNA=false
                if doIncreaseRank
                    nextRank+=rankIncrement
                    rankIncrement=0
                    doIncreaseRank=false
                end
            end
            reducedArrayIndex=reducedArraySortedIndices[groupIndex]
            if !firstFound && lastFound != reducedArraySorted[groupIndex]
                nextRank = NormalizeQuantiles.setRank!(group,nextRank,offset,tiesMethod,result,resultMatrix,rankMatrix)
                group = [naCounts[reducedArrayIndex]+reducedArrayIndex+offset]
                if doIncreaseRank
                    nextRank+=rankIncrement
                    rankIncrement=0
                    doIncreaseRank=false
                end
            else
                push!(group,naCounts[reducedArrayIndex]+reducedArrayIndex+offset)
            end
            lastFound = reducedArraySorted[groupIndex]
            firstFound = false
            groupIndex+=1
        else
            if naIncreasesRank
                doIncreaseRank=true
                rankIncrement+=1
            end
            if firstFound
                headingNA=true
            end
        end
    end
    NormalizeQuantiles.setRank!(group,nextRank,offset,tiesMethod,result,resultMatrix,rankMatrix)
    (result,rankMatrix)
end 

function setRank!(group::Array{Int},nextRank::Int,offset::Int,tiesMethod::qnTiesMethods,result::Array{Union{Missing,Int}},resultMatrix::Bool,rankMatrix::Dict{Int,Array{Int}})
    ranksCount=length(group)
    ranks=nextRank:(nextRank+ranksCount-1)
    if tiesMethod==tmMin
        minRank=minimum(ranks)
        result[group.-offset].=minRank
        if resultMatrix
            rankMatrix[minRank]=group
        end
        nextRank+=1
    elseif tiesMethod==tmMax
        maxRank=maximum(ranks)
        result[group.-offset].=maxRank
        if resultMatrix
            rankMatrix[maxRank]=group
        end
        nextRank=maxRank+1
    elseif tiesMethod==tmOrder
        maxRank=maximum(ranks)
        result[group.-offset]=ranks
        if resultMatrix
            for rankIndex in 1:length(ranks)
                rankMatrix[ranks[rankIndex]]=[group[rankIndex]]
            end
        end
        nextRank=maxRank+1
    elseif tiesMethod==tmReverse
        maxRank=maximum(ranks)
        ranks=reverse(ranks,dims=1)
        result[group.-offset]=ranks
        if resultMatrix
            for rankIndex in 1:length(ranks)
                rankMatrix[ranks[rankIndex]]=[group[rankIndex]]
            end
        end
        nextRank=maxRank+1
    elseif tiesMethod==tmRandom
        maxRank=maximum(ranks)
        ranks=ranks[randperm(ranksCount)]
        result[group.-offset]=ranks
        if resultMatrix
            for rankIndex in 1:length(ranks)
                rankMatrix[ranks[rankIndex]]=[group[rankIndex]]
            end
        end
        nextRank=maxRank+1
    elseif tiesMethod==tmAverage
        rankMean=round(Int,mean(ranks))
        result[group.-offset].=rankMean
        if resultMatrix
            rankMatrix[rankMean]=group
        end
        nextRank=rankMean+1
    end
    nextRank
end

end # module
